MiniAOD in phase2

gpetruc opened the pull request:
Minimal set of changes to get miniAOD production running in the Phase2 era.

Mainly it requires to reconfigure reducedEgamma to ignore ES RecHits, and to switch off the extra electron and photon id stuff, which otherwise is also crashing due to ES RecHits.

Tested in the Phase2C1 scenario (wf 10424.0) in `CMSSW_8_1_X_2016-08-24-1100`, and enabled in the `RecoFullGlobal` chain (though this last change can be backed off easily)

For making the miniAOD_tools customizers era-aware I used the cff + `makeProcessModifier` approach, I can change it when we settle on the best one.

There's some random complaints from tau id, which probably doesn't cope with the updated tracker coverage.

@arizzi @kpedro88 @bendavid @lgray

gpetruc committed:
Get reducedEgamma working with no ES rechits in Phase2

gpetruc committed:
Switch on miniAOD for 2023 workflows that run RecoFullGlobal

cmsbuild commented:
A new Pull Request was created by  @gpetruc (Giovanni Petrucciani) for CMSSW_8_1_X.

It involves the following packages:

Configuration/PyReleaseValidation
PhysicsTools/PatAlgos
RecoEgamma/EgammaPhotonProducers


@cvuosalo, @fabozzi, @cmsbuild, @srimanob, @montjj, @slava77, @hengne, @davidlange6 can you please review it and eventually sign? Thanks.
@ghellwig, @mmarionncern, @Sam-Harper, @rafaellopesdesa, @imarches, @ahinzmann, @rappoccio, @acaudron, @TaiSakuma, @lgray, @Martin-Grunewald, @jdolen, @nhanvtran, @JyothsnaKomaragiri, @schoef, @ferencek, @mverzett, @mariadalfonso, @pvmulder this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are list here cms-sw/cmssw#13028

bendavid commented:
Changes to reducedEgamma look reasonable to me.

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -1,8 +1,11 @@
+
 
 def miniAOD_customizeCommon(process):
+    process.load("PhysicsTools.PatAlgos.slimming.era_customizers")
```
can this be made less circular?

gpetruc replied:
I will improve it when I'll understand the more proper way to properly integrate eras and customises ( @Dr15Jones )

slava77 replied:
you can at least put miniAOD_forPhase2 in a separate file, maybe in the era_customizers itself
right now from file A (this file) you load file B (era_customizers), which loads elements of A

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -322,3 +325,13 @@ def miniAOD_customizeAllMC(process):
+    process.patPhotons.addPhotonID = False 
+    process.patPhotons.photonIDSources = cms.PSet()
+    process.slimmedElectrons.modifyElectrons = False
+    process.slimmedPhotons.modifyPhotons = False
```
the .modify\* changes should be moved to  the slimmedPhotons_cfi and slimmedElectrons_cfi

slava77 replied:
the modifyPhotons are not changed in the customize.
So, the order modify/customize shouldn't  matter

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -322,3 +325,13 @@ def miniAOD_customizeAllMC(process):
+    process.patElectrons.addElectronID = False
+    process.patElectrons.electronIDSources = cms.PSet()
+    process.patPhotons.addPhotonID = False 
+    process.patPhotons.photonIDSources = cms.PSet()
```
isn't setting of addPhotonID and addElectronID to False enough? 
This setting should be done in patElectrons_cfi and patPhotons_cfi

gpetruc replied:
that will work but only if  we decide to change the default PAT behaviour and not only the MiniAOD behaviour (and may have to be changed anyway as soon as we get any id for phase2)

slava77 replied:
do you mean that default PAT works in phase2 out of the box (that's why the miniAOD settings have to be done in miniAOD-only mode a.k.a. customize) ?

gpetruc replied:
all miniAOD settings are always done in a customize (since ever) - I'd like to change this, but I think it has to be done for all eras and not phase2, and I don't know if 81X is the time for doing it.

slava77 replied:
Just to be clear: I want all the edits here (4 of them only : addElectronID addPhotonID, modifyElectrons, and modifyPhotons) be moved to their respective cfi files.
None of these are changed by the fact that these are miniAOD or PAT.

and the setting of electronIDSources and photonIDSources does not change the behavior and can be removed (after the boolean switches are applied)

gpetruc replied:
What I'm trying to explain is that this change would make things nicer only today when there's no valid e/g id or modification for phase2 neither in miniAOD nor in PAT.
When at any moment there will be a e/g id or modification for phase2, then we'll still face the same problem, as just disabling the features in phase2 will no longer work.
So, i'd rather get a real solution to the issue, either by having customizes become better aware of the era or by avoiding the customizes, and in the meantime keep a `miniAOD_forPhase2` function which is a clear placeholder for a better solution (instead of scattering the temporary changes in 4 different files)

slava77 replied:
design of eras is to apply modifications in the cfi files where parameters are defined.
It looks like you are planning to violate it from the start.

customie applies when customize method is called.
era modifiers are applied when corresponding file that defines the era is loaded.

gpetruc replied:
I am trying to follow the spirit of the design: now the configuration for the modifications and IDs for standard run2 is given in the customize, so I'm having the era change the customise (to the extent that it is possible) and have that change defined where the customize is defined.

Dr15Jones replied:
Is the underlying difficulty that miniAOD and PAT are using the same module declarations but want different behaviors for Phase2? What about making miniAOD an era? I believe that is what we did for fast sim. Then there is no need to worry about running the era and a customize together since there will only be the era.

slava77 replied:
+1 for making a miniAOD era

given how PAT configs are setup, it looks like all era changes will still be stuck in a master file

lgray commented:
@gpetruc Can you check this works in HGCal with #15598 and wf11224.0?

gpetruc commented:
@lgray: at a first attempt, this plus #15598 crashes at some ECAL MET filters.
will see if I can get it running (without breaking it for the non-HGCal, for the moment), and if yes update this PR.

lgray commented:
Thanks for taking a look. I am not surprised there are more crashes lurking about.

gpetruc commented:
It seems to me that to get it running with #15598 I need to disable four MET filers and the MET miniAOD DQM: https://github.com/gpetruc/cmssw/compare/miniAOD_phase2...gpetruc:miniAOD_phase2_hgcalCompat
* the HBHE filters need the hcalnoise which is not in phase2 (see 6de71966ee298d047c9b7edf6ae96800673a74f3 from #15393), and the MET DQM needs the filters
* the ECal filters crash.

While disabling the MET filters is probably ok, as they're unlikely to be correct for the phase2 detector, it would still be better to track down why the filters crash (especially the EcalDeadCellTriggerPrimitiveFilter one that crashes in the beginRun, for which I did the ugly workaround of replacing it with a dummy HLTBool module), and also to recover the MET DQM.

lgray commented:
Thanks! I'll take these and see if I can toss you appropriately modified
ECAL filters, it is likely they're just looking for the endcap or preshower
geometries and they aren't there.

On Fri, Aug 26, 2016 at 11:05 AM, Giovanni Petrucciani <
notifications@github.com> wrote:

> It seems to me that to get it running with #15598
> <https://github.com/cms-sw/cmssw/pull/15598> I need to disable four MET
> filers and the MET miniAOD DQM: gpetruc@miniAOD_phase2...
> gpetruc:miniAOD_phase2_hgcalCompat
> <https://github.com/gpetruc/cmssw/compare/miniAOD_phase2...gpetruc:miniAOD_phase2_hgcalCompat>
>
>    - the HBHE filters need the hcalnoise which is not in phase2 (see
>    6de7196
>    <https://github.com/cms-sw/cmssw/commit/6de71966ee298d047c9b7edf6ae96800673a74f3>
>    from #15393 <https://github.com/cms-sw/cmssw/pull/15393>), and the MET
>    DQM needs the filters
>    - the ECal filters crash.
>
> While disabling the MET filters is probably ok, as they're unlikely to be
> correct for the phase2 detector, it would still be better to track down why
> the filters crash (especially the EcalDeadCellTriggerPrimitiveFilter one
> that crashes in the beginRun, for which I did the ugly workaround of
> replacing it with a dummy HLTBool module), and also to recover the MET DQM.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/15615#issuecomment-242777540>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBMObCnDBwvvmGIej55IPZfYJVgX16qks5qjw7AgaJpZM4Jt8Qh>
> .
>

Dr15Jones commented:
@makortel You've had probably the most experience dealing with Modifier system. Any suggestions?

makortel commented:
For the problem at hand, since the entry point to MiniAOD configuration is a customize function, I would use `Modifier.isChosen()` (which is safe when dealing with an explicit `Process` object), i.e. something along
```
def miniAOD_customizeCommon(process):
    if eras.phase2_common.isChosen():
        miniAOD_forPhase2(process)
    ...
```
No `ProcessModifier`, no circularities.

I didn't quite understand what exactly a "MiniAOD era" would solve.

For longer term, I would (also? *) consider making MiniAOD the default "PAT application" (i.e. making the PAT module default configuration parameters to correspond those of MiniAOD). I agree it would need a lot of effort (and might not work in the end), but I believe it is the best alternative for "customize + `isChosen()`".

(*) maybe @gpetruc had already this in mind in https://github.com/cms-sw/cmssw/pull/15615#discussion_r76412883?

Dr15Jones commented:
To replace the need for `isChosen()` which is deprecated, I've added a `Process.isUsingModifier` #15665 
For context, my understanding was Modifiers are intended to allow us to do away with (at least as much as possible) the use of customise functions in our production system. Therefore the two were not really intended to work together.

makortel commented:
@Dr15Jones `Process.isUsingModifier()` looks indeed much better. With it my example becomes
```
def miniAOD_customizeCommon(process):
    if process.isUsingModifier(eras.phase2_common):
        miniAOD_forPhase2(process)
    ...
```

gpetruc commented:
Okay, will update the PR with that, and possibly fix a few other things and
make it already forward-compatible with lindsey's one

Giovanni

Il 31-ago-2016 09:24, "Matti Kortelainen" <notifications@github.com> ha
scritto:

> @Dr15Jones <https://github.com/Dr15Jones> Process.isUsingModifier() looks
> indeed much better. With it my example becomes
>
> def miniAOD_customizeCommon(process):
>     if process.isUsingModifier(eras.phase2_common):
>         miniAOD_forPhase2(process)
>     ...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/15615#issuecomment-243680521>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEbbRx4EqqqO18s7NrSxc5NgSJw7bBZwks5qlSwagaJpZM4Jt8Qh>
> .
>

gpetruc committed:
photonIDValueMapProducer: don't use ES in phase2

gpetruc committed:
EcalDeadCellTriggerPrimitiveFilter: don't use EE in Phase2/HGCal era (c++ code change also needed to avoid crash in beginRun)

gpetruc committed:
Disable MET filters that don't work in current Phase2/HGCal era (only when that era is selected, of course)

gpetruc committed:
Take out MET MiniAOD DQM in Phase2/HGCal era, for which hcalnoise is missing

cmsbuild commented:
Pull request #15615 was updated. @vanbesien, @cvuosalo, @dmitrijus, @cmsbuild, @srimanob, @montjj, @slava77, @fabozzi, @hengne, @davidlange6 can you please check and sign again.

gpetruc commented:
Cleaned up and rebased on the latest 81X IB that has also Lindsey's changes in.
I got it to work without any if statement by fixing the electron ID code rather than disabling it.
It still creates some empty cms.Paths for MET filters that are disabled (I'll try to find a cleaner solution to it later)

slava77 commented:
@cmsbuild please test

lgray commented:
Thanks for patching up the electron ID! Good that it doesn't crash.

On Wed, Aug 31, 2016 at 11:57 AM, Giovanni Petrucciani <
notifications@github.com> wrote:

> Cleaned up and rebased on the latest 81X IB that has also Lindsey's
> changes in.
> I got it to work without any if statement by fixing the electron ID code
> rather than disabling it.
> It still creates some empty cms.Paths for MET filters that are disabled
> (I'll try to find a cleaner solution to it later)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/15615#issuecomment-243829422>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABBMOd-kPan6UHO0nfKzosn_jpI0_xfdks5qlbJcgaJpZM4Jt8Qh>
> .
>

cmsbuild commented:
The tests are being triggered in jenkins.

gpetruc commented:
Tested locally that both 10424.0 (Phase2C1) and 11224.0 (Phase2C2) run
without errors.

Giovanni

On Wed, Aug 31, 2016 at 7:06 PM, cmsbuild <notifications@github.com> wrote:

> The tests are being triggered in jenkins.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/15615#issuecomment-243832053>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AEbbRyDILx2bWNVT55xJjo4LQ88Sqe3uks5qlbSAgaJpZM4Jt8Qh>
> .
>

cmsbuild commented:
+1
Tested at: e46ef8b3b3496cfeea033a1f03626c0b9211313d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15615/14873/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-15615/14873/summary.html

cvuosalo commented:
+1

For #15615 e46ef8b3b3496cfeea033a1f03626c0b9211313d

Adding Mini-AOD production to Phase 2.

The code changes are satisfactory, and Jenkins tests against baseline CMSSW_8_1_X_2016-08-31-1100 show no significant differences. The Jenkins Phase 2 workflows now produce Mini-AOD output.

dmitrijus commented:
+1
